sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/TextBasedFileScan.scala

```diff
@@ -33,14 +34,16 @@ abstract class TextBasedFileScan(
     readPartitionSchema: StructType,
     options: CaseInsensitiveStringMap)
   extends FileScan(sparkSession, fileIndex, readDataSchema, readPartitionSchema) {
-  private var codecFactory: CompressionCodecFactory = _
+  @transient private lazy val codecFactory: CompressionCodecFactory = new CompressionCodecFactory(
+    sparkSession.sessionState.newHadoopConfWithOptions(options.asScala.toMap))
 
-  override def isSplitable(path: Path): Boolean = {
-    if (codecFactory == null) {
-      codecFactory = new CompressionCodecFactory(
-        sparkSession.sessionState.newHadoopConfWithOptions(options.asScala.toMap))
+  override def isSplitable(path: Path): Boolean = Utils.isFileSplittable(path, codecFactory)
+
+  override def getFileUnSplittableReason(path: Path): String = {
+    if (!isSplitable(path)) {
+      "the file is compressed by unsplittable compression codec"
+    } else {
+      null
```

Comment:
when will we return null here?

Reply:
If path is splittable, return null. (because this method is to get path "unsplittable" reason)

Reply:
but we won't call `getFileUnSplittableReason` if the file is splittable, or do I miss something?

Reply:
At least we should return `undefined` here.

Reply:
What about add an `require` inside method `getFileUnSplittableReason`
`require(!isSplitable(path), "file is splittable.")`

Reply:
sure

Reply:
I update return type to be`Option[String]` (when file is splittable return `None`).
So that it is more fit with scala style.